server-spi-private/src/main/java/org/keycloak/vault/VaultProvider.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.vault;
+
+import org.keycloak.provider.Provider;
+
+import java.util.Optional;
+
+/**
+ * Provider interface for a vault. The only purpose of a vault is retrieval of secrets.
+ */
+public interface VaultProvider extends Provider {
+
+    /**
+     * Retrieves a secret from vault. The implementation should respect
+     * at least the realm ID to separate the secrets within the vault.
+     * If the secret is retrieved successfully, it is returned;
+     * otherwise this method results into an empty {@link java.util.Optional}.
+     *
+     * @param vaultSecretId Identifier of the secret. It corresponds to the value
+     *        entered by user in the respective configuration, which in turn
+     *        is obtained from the vault when storing the secret.
+     *
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(String vaultSecretId);
```

Comment:
So the other idea is to put `WeakReference<String>` here.

Reply:
Vault generally returns a bit sequence, which in Java the closest object is byte array. `String` is interpretation of this byte sequence. Hence String would only be obtained from that byte array in the transcriber object.